hal/architecture/MyHwNRF5.cpp

```diff
@@ -0,0 +1,436 @@
+
+void hwSleepPrepare(unsigned long ms)
+{
+	// Enable low power sleep mode
+	NRF_POWER->TASKS_LOWPWR = 1;
+
+	if (ms > 0) {
+		// Configure RTC
+#ifdef NRF51
+		MY_HW_RTC->POWER = 1;
+#endif
+		// Reset RTC
+		MY_HW_RTC->TASKS_CLEAR = 1;
+
+		// Calculate sleep time
+		// 8 Hz -> max 582.542 hours sleep.
+		MY_HW_RTC->PRESCALER = 4095;
+		// Set compare register to 1/125ms + 2 to garantee event triggering
+		MY_HW_RTC->CC[0] = (ms / 125) + 2;
+
+		MY_HW_RTC->INTENSET = RTC_INTENSET_COMPARE0_Msk;
+		MY_HW_RTC->EVTENSET = RTC_EVTENSET_COMPARE0_Msk;
+		MY_HW_RTC->EVENTS_COMPARE[0] = 0;
+		nrf5_rtc_event_triggered = false;
+		MY_HW_RTC->TASKS_START = 1;
+		NVIC_SetPriority(MY_HW_RTC_IRQN, 15);
+		NVIC_ClearPendingIRQ(MY_HW_RTC_IRQN);
+		NVIC_EnableIRQ(MY_HW_RTC_IRQN);
+	} else {
+		NRF_RTC1->TASKS_STOP = 1;
+	}
+
+	// Stop HFCLK
+	nrf5_pwr_hfclk = NRF_CLOCK->EVENTS_HFCLKSTARTED;
+	NRF_CLOCK->TASKS_HFCLKSTOP = 1;
+
+	// Idle serial device
+#ifndef MY_DISABLED_SERIAL
+	NRF_UART0->TASKS_STOPRX = 1;
+	NRF_UART0->TASKS_STOPTX = 1;
+	NRF_UART0->TASKS_SUSPEND = 1;
+#endif
+
+	// Clear NVRAM log, if needed and a time frame of 4 seconds available
+	if (ms > 40000) {
+		// preserve some bytes for writing to NVRAM
+		NVRAM.clean_up(32);
+	}
+}
+
+void hwSleepEnd(unsigned long ms)
+{
+	// Start HFCLK
+	if (nrf5_pwr_hfclk) {
+		NRF_CLOCK->EVENTS_HFCLKSTARTED = 0;
+		NRF_CLOCK->TASKS_HFCLKSTART = 1;
+		while (NRF_CLOCK->EVENTS_HFCLKSTARTED == 0)
+			;
+		// Enable low latency sleep mode
+		NRF_POWER->TASKS_CONSTLAT = 1;
+	}
+
+	if (ms > 0) {
+		// Stop RTC
+#ifdef NRF51
+		MY_HW_RTC->POWER = 0;
+#endif
+		MY_HW_RTC->INTENCLR = RTC_INTENSET_COMPARE0_Msk;
+		MY_HW_RTC->EVTENCLR = RTC_EVTENSET_COMPARE0_Msk;
+		MY_HW_RTC->TASKS_STOP = 1;
+		NVIC_DisableIRQ(MY_HW_RTC_IRQN);
+	} else {
+		// Start Arduino RTC for millis()
+		NRF_RTC1->TASKS_START = 1;
+	}
+
+	// Start serial device
+#ifndef MY_DISABLED_SERIAL
+	NRF_UART0->TASKS_STARTRX = 1;
+	NRF_UART0->TASKS_STARTTX = 1;
+#endif
+}
+
+// Halt CPU until next interrupt event
+inline void hwWaitForInterrupt()
+{
+	__DSB();
+	__WFI();
+}
+
+// Sleep in System ON mode
+inline void hwSleep()
+{
+	__WFE();
+	__SEV();
+	__WFE();
+}
+
+int8_t hwSleep(unsigned long ms)
+{
+	hwSleepPrepare(ms);
+	while (nrf5_rtc_event_triggered == false) {
+		hwSleep();
+	}
+	hwSleepEnd(ms);
+	return MY_WAKE_UP_BY_TIMER;
+}
+
+int8_t hwSleep(uint8_t interrupt, uint8_t mode, unsigned long ms)
+{
+	return hwSleep(interrupt, mode, INVALID_INTERRUPT_NUM, 0u, ms);
+}
+
+int8_t hwSleep(uint8_t interrupt1, uint8_t mode1, uint8_t interrupt2,
+               uint8_t mode2, unsigned long ms)
+{
+	// Disable interrupts until going to sleep, otherwise interrupts occurring
+	// between attachInterrupt()
+	// and sleep might cause the MCU to not wakeup from sleep as interrupt has
+	// already be handled!
+	MY_CRITICAL_SECTION {
+		// Fix cppcheck "(style) Variable '__savePriMask' is assigned a value that is never used."
+		(void)__savePriMask;
```

Comment:
I don't have any better idea about getting rid this message. This needs a fix in MyHwNRF5.h and MyHwSAMD.h